diff -pruN a/dlls/shell32/shellpath.c b/dlls/shell32/shellpath.c
--- a/dlls/shell32/shellpath.c	2021-07-28 02:43:14.362356984 -0300
+++ b/dlls/shell32/shellpath.c	2021-07-28 04:40:07.314106181 -0300
@@ -4092,75 +4092,18 @@ static void _SHCreateMyDocumentsSubDirs(
  */
 static void _SHCreateMyDocumentsSymbolicLink(const UINT * aidsMyStuff, const UINT aids_num)
 {
-    static const char * const xdg_dirs[] = { "DOCUMENTS" };
-    static const unsigned int num = ARRAY_SIZE(xdg_dirs);
     char szPersonalTarget[FILENAME_MAX], *pszPersonal;
-    struct stat statFolder;
-    const char *pszHome;
-    char ** xdg_results;
 
     /* Get the unix path of 'My Documents'. */
     pszPersonal = _SHGetFolderUnixPath(CSIDL_PERSONAL|CSIDL_FLAG_DONT_VERIFY);
     if (!pszPersonal) return;
 
-    _SHGetXDGUserDirs(xdg_dirs, num, &xdg_results);
-
-    pszHome = getenv("HOME");
-    if (pszHome && !stat(pszHome, &statFolder) && S_ISDIR(statFolder.st_mode))
-    {
-        while (1)
-        {
-            /* Check if there's already a Wine-specific 'My Documents' folder */
-            strcpy(szPersonalTarget, pszHome);
-            if (_SHAppendToUnixPath(szPersonalTarget, MAKEINTRESOURCEW(IDS_PERSONAL)) &&
-                !stat(szPersonalTarget, &statFolder) && S_ISDIR(statFolder.st_mode))
-            {
-                /* '$HOME/My Documents' exists. Create subfolders for
-                 * 'My Pictures', 'My Videos', 'My Music' etc. or fail silently
-                 * if they already exist.
-                 */
-                _SHCreateMyDocumentsSubDirs(aidsMyStuff, aids_num, szPersonalTarget);
-                break;
-            }
-
-            /* Try to point to the XDG Documents folder */
-            if (xdg_results && xdg_results[0] &&
-               !stat(xdg_results[0], &statFolder) &&
-               S_ISDIR(statFolder.st_mode))
-            {
-                strcpy(szPersonalTarget, xdg_results[0]);
-                break;
-            }
-
-            /* Or the hardcoded / OS X Documents folder */
-            strcpy(szPersonalTarget, pszHome);
-            if (_SHAppendToUnixPath(szPersonalTarget, DocumentsW) &&
-               !stat(szPersonalTarget, &statFolder) &&
-               S_ISDIR(statFolder.st_mode))
-                break;
-
-            /* As a last resort point to $HOME. */
-            strcpy(szPersonalTarget, pszHome);
-            break;
-        }
-
-        /* Create symbolic link to 'My Documents' or fail silently if a directory
-         * or symlink exists. */
-        symlink(szPersonalTarget, pszPersonal);
-    }
-    else
-    {
-        /* '$HOME' doesn't exist. Create subdirs for 'My Pictures', 'My Videos',
-         * 'My Music' etc. in '%USERPROFILE%\My Documents' or fail silently if
-         * they already exist. */
-        pszHome = NULL;
-        strcpy(szPersonalTarget, pszPersonal);
-        _SHCreateMyDocumentsSubDirs(aidsMyStuff, aids_num, szPersonalTarget);
-    }
+    /* Always default to '%USERPROFILE%\My Documents' to create subdirs for 'My Pictures', 'My Videos',
+     * 'My Music' etc. or fail silently if they already exist. */
+    strcpy(szPersonalTarget, pszPersonal);
+    _SHCreateMyDocumentsSubDirs(aidsMyStuff, aids_num, szPersonalTarget);
 
     heap_free(pszPersonal);
-
-    _SHFreeXDGUserDirs(num, xdg_results);
 }
 
 /******************************************************************************
